software/py/vanilla_stats_parser.py

```diff
@@ -0,0 +1,640 @@
+        for op in self.all_ops_list:
+          manycore_stat[op] += tile_stat[y][x][op]
+
+    return manycore_stat
+ 
+
+
+  # prints all four types of stats, timing, instruction,
+  # miss and stall for the entire manycore 
+  def print_manycore_stats_all(self):
+    stats_path = os.getcwd() + "/stats/"
+    if not os.path.exists(stats_path):
+      os.mkdir(stats_path)
+    manycore_stats_file = open( (stats_path + "manycore_stats.log"), "w")
+    self.__print_manycore_stats_tile_group_timing(manycore_stats_file)
+    self.__print_manycore_stats_tile_timing(manycore_stats_file)
+    self.__print_manycore_stats_miss(manycore_stats_file)
+    self.__print_manycore_stats_stall(manycore_stats_file)
+    self.__print_manycore_stats_instr(manycore_stats_file)
+    manycore_stats_file.close()
+
+  # prints all four types of stats, timing, instruction,
+  # miss and stall for each tile group in a separate file  
+  def print_per_tile_group_stats_all(self):
+    stats_path = os.getcwd() + "/stats/tile_group/"
+    if not os.path.exists(stats_path):
+      os.mkdir(stats_path)
+    for tg_id in range(self.num_tile_groups):
+      stat_file = open( (stats_path + "tile_group_" + str(tg_id) + "_stats.log"), "w")
+      self.__print_per_tile_group_stats_timing(tg_id, stat_file)
+      self.__print_per_tile_group_stats_miss(tg_id, stat_file)
+      self.__print_per_tile_group_stats_stall(tg_id, stat_file)
+      self.__print_per_tile_group_stats_instr(tg_id, stat_file)
+      stat_file.close()
+
+
+
+  # prints all four types of stats, timing, instruction,
+  # miss and stall for each tile in a separate file  
+  def print_per_tile_stats_all(self):
+    stats_path = os.getcwd() + "/stats/tile/"
+    if not os.path.exists(stats_path):
+      os.mkdir(stats_path)
+    for y in range(self.manycore_dim_y):
+      for x in range(self.manycore_dim_x):
+        stat_file = open( (stats_path + "tile_" + str(y) + "_" + str(x) + "_stats.log"), "w")
+        self.__print_per_tile_stats_timing(y, x, stat_file)
+        self.__print_per_tile_stats_miss(y, x, stat_file)
+        self.__print_per_tile_stats_stall(y, x, stat_file)
+        self.__print_per_tile_stats_instr(y, x, stat_file)
+        stat_file.close()
+
+
+  # main public method
+  # default stats generator
+  def generate_stats(self, input_file):
+    self.traces = []
+    with open(input_file) as f:
+      csv_reader = csv.DictReader (f, delimiter=",")
+
+      # Generate list of instruction/miss/stall types
+      header = csv_reader.fieldnames
+      for item in header:
+        if (item.startswith('instr_')):
+          if (not item == 'instr_total'):
+            self.instr_list += [item]
+        elif (item.startswith('miss_')):
+          self.miss_list += [item]
+        elif (item.startswith('stall_')):
+          self.stalls_list += [item]
+        else:
+          self.stats_list += [item]
+      self.all_ops_list = self.stats_list + self.instr_list + self.miss_list + self.stalls_list
```

Comment:
To be honest, I'd do all of this in __init__, since it's O(1), and done every time. Then you don't have to worry about incorrect object initialization/object-function-call ordering

Reply:
The entire method is called inside the init() though, do you mean just scrap the function and move it's body inside the init()? that'd be a bit dirty wouldn't it? I can make the method private though, to make sure nobody calls it from outside to disrupt the function call ordering, what do you think?

Reply:
I mean, parsing the field names. 

To summarize what I find a weird is that your main "public" method is called in init and does initialization. Those are two distinct operations (initialization, and processing). It sets instance variables (self.tile_stat, self.tile_group_stat), and returns the same values. 

I think it should be broken down to be less monolithic - the functionality may be the same, but it'll be easier to understand what initialization means. 

For example, one way to break this down is to initialize all variables inside of __init__ (you do this now), and then call __parse_ops_list (which returns the parsed CSV header) and returns (not sets) the list of ops (which are set inside of __init__). Then you call __parse stats/__generate_tile stats, which returns (not sets) the tile stats, etc...


To summarize what's weird, is that this function sets or modifies instance variables, but isn't an init method, and isn't a method prepended with __init (to indicate initialization)